src/lsm/tree.zig

```diff
@@ -227,32 +234,166 @@ pub fn TreeType(comptime Table: type) type {
                 if (tree.table_mutable.get(key) orelse
                     tree.value_cache.?.getKeyPtr(tombstone_from_key(key))) |value|
                 {
-                    callback(unwrap_tombstone(value));
+                    callback(context, unwrap_tombstone(value));
                     return;
                 }
             }
 
-            // Hash the key to the fingerprint only once and reuse for all bloom filter checks.
             if (!tree.table_immutable.free and tree.table_immutable.snapshot_min < snapshot) {
                 if (tree.table_immutable.get(key)) |value| {
-                    callback(unwrap_tombstone(value));
+                    callback(context, unwrap_tombstone(value));
                     return;
                 }
             }
 
-            var it = tree.manifest.lookup(snapshot, key);
-            if (it.next()) |info| {
-                assert(info.visible(snapshot));
-                assert(compare_keys(key, info.key_min) != .lt);
-                assert(compare_keys(key, info.key_max) != .gt);
-
-                // TODO
-            } else {
-                callback(null);
-                return;
+            var index_block_count: u8 = 0;
+            var index_block_addresses: [config.lsm_levels]u64 = undefined;
+            var index_block_checksums: [config.lsm_levels]u128 = undefined;
+            {
+                var it = tree.manifest.lookup(snapshot, key);
+                while (it.next()) |table| : (index_block_count += 1) {
+                    assert(table.visible(snapshot));
+                    assert(compare_keys(table.key_min, key) != .gt);
+                    assert(compare_keys(table.key_max, key) != .lt);
+
+                    index_block_addresses[index_block_count] = table.address;
+                    index_block_checksums[index_block_count] = table.checksum;
+                }
             }
+
```

Comment:
What do we expect if `index_block_count==0` at the end of the loop?

I think we could return early here without the cost of computing the fingerprint if we know that no table in any level contains the key.

Reply:
Good catch, this is actually a bug introduced in the refactor to do iteration up front. Right now we'll end up using an undefined address/checksum to try and read an index block from disk. With this change and the `index_block_count > 0` assert we'll have the bug fixed and be protected against re-introducing it.